# Using Spotify's audio analysis to find the starting point of Nicki Minaj's verse in Side to Side by Ariana Grande

### Step 1: To begin I used pip to install spotipy

In [3]:
pip install spotipy

Note: you may need to restart the kernel to use updated packages.


'c:\users\rutho\pycharmprojects\cfg-python\cfg' is not recognized as an internal or external command,
operable program or batch file.


### Step 2: Then I imported spotipy, and authorised it using my client creditionals.

#### I imported pandas to make a dataframe to make the data more manageable

In [2]:
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials

sp = spotipy.Spotify(auth_manager=SpotifyClientCredentials(client_id="56ab3cc7108646a98b183042b661114b",
                                                           client_secret="17fda25e18164fbcb133a6621e453242"))

track_id = '6j4w8O8gEseBUZTOo8vYMK'
feat_results = sp.audio_analysis(track_id)

import pandas as pd


df = pd.DataFrame(feat_results["segments"])
df

start  duration  confidence  loudness_start  loudness_max_time  \
0      0.00000   0.18567       0.000         -60.000            0.00000   
1      0.18567   0.18594       1.000         -60.000            0.11628   
2      0.37161   0.13778       0.607         -31.267            0.11020   
3      0.50939   0.33737       0.398         -25.026            0.11847   
4      0.84676   0.07583       0.410         -25.031            0.03583   
..         ...       ...         ...             ...                ...   
571  173.97420   0.16961       0.590         -35.042            0.04772   
572  174.14381   0.09315       0.352         -35.980            0.01355   
573  174.23695   0.35991       0.180         -36.758            0.03734   
574  174.59686   0.50975       0.312         -42.060            0.05910   
575  175.10661   2.92005       0.077         -50.044            0.00910   

     loudness_max  loudness_end  \
0         -60.000           0.0   
1         -31.955           0.0   
2         -25.128           0.0   
3         -19.309           0.0   
4         -21.423           0.0   
..            ...           ...   
571       -29.462           0.0   
572       -30.837           0.0   
573       -32.605           0.0   
574       -38.306           0.0   
575       -48.798         -60.0   

                                               pitches  \
0    [1.0, 0.996, 0.637, 0.61, 0.59, 0.59, 0.58, 0....   
1    [1.0, 0.06, 0.035, 0.055, 0.428, 0.047, 0.081,...   
2    [0.711, 0.09, 0.025, 0.086, 1.0, 0.052, 0.034,...   
3    [1.0, 0.074, 0.031, 0.034, 0.465, 0.24, 0.072,...   
4    [0.831, 0.274, 0.033, 0.088, 0.411, 0.088, 0.0...   
..                                                 ...   
571  [0.292, 0.165, 0.255, 0.184, 1.0, 0.54, 0.191,...   
572  [0.202, 0.196, 0.294, 1.0, 0.731, 0.308, 0.11,...   
573  [0.181, 0.197, 0.092, 0.05, 1.0, 0.118, 0.074,...   
574  [0.556, 0.64, 0.59, 0.589, 1.0, 0.922, 0.498, ...   
575  [0.108, 0.107, 0.125, 0.166, 0.497, 0.289, 0.1...   

                                                timbre  
0    [0.0, 171.13, 9.469, -28.48, 57.491, -50.067, ...  
1    [24.24, 2.363, 73.003, -54.269, 47.801, 7.108,...  
2    [31.592, -12.09, 81.929, -51.673, 35.334, -24....  
3    [37.29, -19.682, 48.519, -25.153, 19.304, -20....  
4    [38.278, 4.486, 74.627, -11.983, 15.321, -33.6...  
..                                                 ...  
571  [27.589, -169.393, 22.967, -2.968, 56.486, -18...  
572  [25.762, -52.051, 36.086, 30.575, 24.852, -42....  
573  [23.614, -125.516, 24.399, 42.905, 58.133, -24...  
574  [16.842, -97.04, 0.66, 40.877, 110.147, -45.83...  
575  [6.039, 54.754, -18.672, 47.654, 104.474, -89....  

[576 rows x 9 columns]

### Step 3: I wanted to access the timbre of the audio since that is unique to everyone regardless of the backing track/beat/instrumental

#### "Vocal Timbre, or as it is described as the quality of that tone utilizing complex overtones, or sound waves, is that unique “something” that gives color and personality to your voice, and how it is recognized."

#### Every voice has its own distinguished timbre

In [3]:
df.drop(columns=['confidence', 'pitches', 'loudness_start', 'loudness_end', 'loudness_max','loudness_max_time','duration',])

start                                             timbre
0      0.00000  [0.0, 171.13, 9.469, -28.48, 57.491, -50.067, ...
1      0.18567  [24.24, 2.363, 73.003, -54.269, 47.801, 7.108,...
2      0.37161  [31.592, -12.09, 81.929, -51.673, 35.334, -24....
3      0.50939  [37.29, -19.682, 48.519, -25.153, 19.304, -20....
4      0.84676  [38.278, 4.486, 74.627, -11.983, 15.321, -33.6...
..         ...                                                ...
571  173.97420  [27.589, -169.393, 22.967, -2.968, 56.486, -18...
572  174.14381  [25.762, -52.051, 36.086, 30.575, 24.852, -42....
573  174.23695  [23.614, -125.516, 24.399, 42.905, 58.133, -24...
574  174.59686  [16.842, -97.04, 0.66, 40.877, 110.147, -45.83...
575  175.10661  [6.039, 54.754, -18.672, 47.654, 104.474, -89....

[576 rows x 2 columns]

### Step 4: The results returned the timbre in the form of a list so I calculated the average using a for loop

In [4]:
for feat_results in feat_results["segments"]:
    
    print(sum(feat_results["timbre"]))


135.61200000000002
97.23200000000003
21.256000000000004
-5.064000000000011
92.59099999999997
32.87100000000001
-28.855999999999987
-137.44100000000003
-65.50699999999999
30.71700000000001
-114.28299999999999
28.745
23.739999999999995
58.849000000000004
96.15300000000002
150.49500000000003
129.91700000000006
129.41199999999998
118.38999999999996
89.52100000000004
137.422
126.48300000000003
105.83999999999995
55.21099999999998
132.643
36.30499999999999
18.375
-15.690999999999995
19.69399999999998
42.41999999999997
23.017999999999976
52.515
-120.46199999999997
-155.35600000000002
-119.32799999999999
-75.983
-87.323
-73.396
-109.49700000000001
-8.13399999999999
-9.433999999999997
23.592999999999975
-44.095000000000006
-42.67700000000001
-4.930000000000006
-41.81800000000001
-3.6340000000000074
18.225999999999996
-36.019000000000005
-56.136999999999986
-12.412999999999995
-35.21999999999999
-31.893
2.813000000000005
49.67400000000002
39.017
-5.4639999999999915
1.5310000000000032
50.96100000

### Step 5:  I calculated the start of Nicki's verse manually by listening and using some calculations 

In [9]:
start_time = 2.14*60
print('Nicki Minajs verse ends at {} seconds'.format(start_time))

end_time = 2.51*60
print('Nicki Minajs verse ends at {} seconds'.format(end_time))


Nicki Minajs verse ends at 128.4 seconds
Nicki Minajs verse ends at 150.6 seconds


### Step 6: I extracted the data I needed manually then exported it to csv

In [7]:
compare_timbre = pd.read_csv("audio_analysis.csv")
compare_timbre

Time_s  full_timbre_Hz  nicki_timbre_Hz  ariana_timbre_Hz
0      0.00000         135.612              0.0             0.000
1      0.18567          97.232              0.0             0.000
2      0.37161          21.256              0.0             0.000
3      0.50939          -5.064              0.0             0.000
4      0.84676          92.591              0.0             0.000
..         ...             ...              ...               ...
571  173.97420         -72.057              0.0           -72.057
572  174.14381         -51.504              0.0           -51.504
573  174.23695         -97.407              0.0           -97.407
574  174.59686          -6.128              0.0            -6.128
575  175.10661         131.474              0.0           131.474

[576 rows x 4 columns]

### Step 7: Then I did some data analysis to find a trend

In [8]:
compare_timbre.describe()

Time_s  full_timbre_Hz  nicki_timbre_Hz  ariana_timbre_Hz
count  576.000000      576.000000       576.000000        576.000000
mean    86.924811       26.518354        -4.564705         29.963021
std     50.930510      127.575400        60.456598        108.672746
min      0.000000     -419.985000      -419.985000       -298.485000
25%     42.099537      -56.293750         0.000000         -6.193750
50%     87.910700       21.943000         0.000000          0.000000
75%    131.863677      104.483750         0.000000         85.561250
max    175.106610      616.812000       404.496000        616.812000

### Step 8: The maximum and mininum values stood out to me - because Timbre values fluctuate a lot I wanted to find the "outliers" 
#### Following this I plotted a graph with this data

<img src="graphproject.png">

### Step 9: I calculated the biggest difference in timbre

In [9]:
(compare_timbre['full_timbre_Hz'] - compare_timbre['nicki_timbre_Hz']).idxmax()

197

### Step 10: Then I found the time stamps greater than or equal to 197 Hz

In [13]:
compare_timbre[compare_timbre['nicki_timbre_Hz'] >= 197]

Time_s  full_timbre_Hz  nicki_timbre_Hz  ariana_timbre_Hz
424  129.02150         273.564          273.564               0.0
425  129.15977         261.227          261.227               0.0
426  129.29501         289.083          289.083               0.0
427  129.82295         328.471          328.471               0.0
453  138.38992         404.496          404.496               0.0
456  139.32450         273.438          273.438               0.0
485  146.96558         267.222          267.222               0.0

In [10]:
compare_timbre.iloc[424,0]

129.0215

#### This timestamp was about 0.6 seconds from the start of nicki's verse that I had calculated prior

In [24]:
timestamp_cal = compare_timbre.iloc[424,0]
print(timestamp_cal)

difference = timestamp_cal - start_time

print(difference)

129.0215
0.6214999999999975


### Step 11: Then I had a look at the minimum and maximum values

In [15]:
compare_timbre.groupby(['full_timbre_Hz']).min()

Time_s  nicki_timbre_Hz  ariana_timbre_Hz
full_timbre_Hz                                              
-419.985        148.03810         -419.985             0.000
-309.655        148.57819         -309.655             0.000
-298.485        160.89362            0.000          -298.485
-283.843         25.71660            0.000          -283.843
-255.910        169.47125            0.000          -255.910
...                   ...              ...               ...
 362.614        121.22585            0.000           362.614
 404.496        138.38992          404.496             0.000
 415.162         34.89392            0.000           415.162
 428.355        121.44603            0.000           428.355
 616.812         61.28295            0.000           616.812

[576 rows x 3 columns]

### Step 12: On average Nicki had a lower timbre and the minimum timbre of the song was recorded during her verse

In [62]:
compare_timbre[compare_timbre['nicki_timbre_Hz'] == -419.985]

Time_s  full_timbre_Hz  nicki_timbre_Hz  ariana_timbre_Hz
490  148.0381        -419.985         -419.985               0.0

In [17]:
compare_timbre.iloc[490,0]

148.0381

#### This was approximately 2.6 seconds shorter than the end time I calculated.

In [33]:
end_time_cal = compare_timbre.iloc[490,0]
end_verse = end_time - end_time_cal
print(end_verse)

2.5619000000000085


### Step 13: The median time I calculated was 1 second off the time when the max timbre recorded during Nicki's verse

In [36]:
median = start_time + end_time
median_time = median/2
print(median_time)

139.5


In [38]:
compare_timbre[compare_timbre['Time_s'] == 138.38992]

Time_s  full_timbre_Hz  nicki_timbre_Hz  ariana_timbre_Hz
453  138.38992         404.496          404.496               0.0

In [37]:
compare_timbre.iloc[453,0]

138.38992

In [39]:
print(median_time - compare_timbre.iloc[453,0])

1.1100800000000106


### In conclusion:

#### 1. The biggest difference in timbre was 197Hz, at 129.02150s - where Nicki's verse starts.

#### 2. The maximum timbre during Nicki's verse was 404.496 Hz - recorded at 138.38992s, in the middle of her verse.

#### 3. The minimum timbre for the whole song was -419.985 Hz at 148.03810s where Nicki Minajs verse ends.

#### I wanted to use this data to create a code for a user but I found it quite difficult since the timbre was in a nested list, and the numbers within the list fluctuated.

### Instead I created a user led spotify search code.

In [8]:
user_artist = input('What artist are you looking for? ')

user_track_no = int(input('How many tracks? '))

results = sp.search(q='{}'.format(user_artist), limit='{}'.format(user_track_no))
for idx, track in enumerate(results['tracks']['items']):

    print(idx, track['name'])


What artist are you looking for? Nicki Minaj
How many tracks? 10
0 Tusa
1 What That Speed Bout!?
2 Whole Lotta Choppas (Remix) [feat. Nicki Minaj]
3 Super Bass
4 Yikes
5 Bottoms Up (feat. Nicki Minaj)
6 FEFE
7 Starships
8 Side To Side
9 MotorSport


# My spotify developer dashboard

<img src="spotifydeveloper.png">

### The end!